### PR TITLE
Add Operators and Nominator Roles Verification

### PR DESCRIPTION
**Pull request by marc-aurele-besner**
## Add Operators and Nominator Roles Verification

This add role verification logic for Operators and Nominators

___

## **Type**
Enhancement, Bug fix

___

## **Description**
- Updated GraphQL queries and types to include checks for multiple roles (farmer, operator, nominator).
- Refactored Discord role components to handle multiple roles and improved the logic for role verification and assignment.
- Enhanced authentication providers to support verification of multiple roles.
- Added utility functions for managing Discord roles for operators and nominators.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>gql.ts</strong><dd><code>Update GraphQL Queries and Functions for Role Checks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/gql/gql.ts
<li>Updated the <code>CheckRole</code> GraphQL query to include checks for operator and <br>nominator roles.<br> <li> Modified the <code>graphql</code> function to use the updated <code>CheckRole</code> query <br>structure.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-838eb3e23d761e38d952104619cc66613a9bf01b3313bbe6923652545f805e7e">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>graphql.ts</strong><dd><code>Update GraphQL Types for Role Verification</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/gql/graphql.ts
<li>Updated the <code>CheckRoleQuery</code> type to include operator and nominator <br>fields.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-bc0e46e46459f8be09a193e520663a980b0e3512600117b0db7772f0d66a1290">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>GetDiscordRoles.tsx</strong><dd><code>Refactor Discord Role Components and Logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/WalletSideKick/GetDiscordRoles.tsx
<li>Refactored <code>GetDiscordRoles</code> component to handle multiple roles and <br>renamed it to <code>SubspaceWalletFlow</code>.<br> <li> Added new <code>DiscordFlow</code> component for Discord connection logic.<br> <li> Updated component to use new GraphQL query for role checks.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-c9fc1acc2edd6e9305ede895667b905fb8513cf39efb6d38d70989989c5a2c3b">+83/-70</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>query.ts</strong><dd><code>Update GraphQL Query for Role Checks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/components/WalletSideKick/query.ts
<li>Renamed <code>QUERY_CHECK_ROLE</code> to <code>QUERY_CHECK_ROLES</code> and updated it to check <br>for multiple roles.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-1e4b2758f6a433a19e644ca2de5175dfd6ca3938db71d34da66ff39fd764eceb">+16/-2</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>discord.ts</strong><dd><code>Enhance Discord Authentication for Multiple Roles</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/utils/auth/providers/discord.ts
<li>Enhanced Discord authentication to handle multiple roles (farmer, <br>operator, nominator).<br> <li> Implemented role verification and assignment based on session data.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-8cadeb7edb2c1f7b85c5d0811be70f986af4c5540c6aaaf3f39c40d39474cf82">+27/-5</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>subspace.ts</strong><dd><code>Update Subspace Provider for Role Verification</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/utils/auth/providers/subspace.ts
<li>Updated Subspace provider to verify multiple roles (farmer, operator, <br>nominator).<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-e42d5e073cf25b6dbbcebef7de8678732ee361cd1e6237872c4c095c97a290f4">+4/-5</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Add Utility Functions for Discord Role Management</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/utils/auth/vcs/discord/index.ts
<li>Added utility functions for operator and nominator role management in <br>Discord.<br>

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-37d9d7de9eedf356367d8391fa27a50c91374466918bc349494a75ae317468e8">+13/-1</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Update Subspace VCS for Multiple Role Checks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

explorer/src/utils/auth/vcs/subspace/index.ts
- Updated to check multiple roles using the new GraphQL query.

</details>
    

  </td>
  <td><a href="https://github.com/subspace/astral/pull/577/files#diff-d33d136e73a33fcad4e2f7197dd3a5fb2b4cd64839eee2a53b129239bd3d0fe1">+20/-2</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by jim-counter** [comment]
We just need to get you the role IDs as everything else should already be in place here @marc-aurele-besner?

File: explorer/.env.sample [hunk 1]
```diff
@@ -19,6 +19,12 @@ DISCORD_BOT_TOKEN="<discord-bot-token>"
 
 # Subspace Discord Server Role ID's
 DISCORD_GUILD_ROLE_ID_FARMER="<discord-farmer-role-id>"
+DISCORD_GUILD_ROLE_ID_OPERATOR="<discord-operator-role-id>"
+DISCORD_GUILD_ROLE_ID_NOMINATOR="<discord-nominator-role-id>"
+
+# GitHub integration for authentication
+GITHUB_CLIENT_ID="<github-client-id>"
```
Comment (EmilFattakhov) [inline explorer/.env.sample:26]: Is this a specific Repo ID we provide here, or we can pick the whole organization? We likely want to give labels for the multiple repo contributors.

File: explorer/gql/gql.ts [hunk 2]
```diff
@@ -48,7 +48,7 @@ const documents = {
...
-    "\n  query CheckRole($subspaceAccount: String!) {\n    isFarmer: rewardEvents(\n      where: { name_eq: \"Rewards.VoteReward\", account: { id_eq: $subspaceAccount } }\n      limit: 1\n    ) {\n      account {\n        id\n      }\n    }\n  }\n": types.CheckRoleDocument,
+    "\n  query CheckRole($subspaceAccount: String!) {\n    farmer: rewardEvents(\n      where: { name_eq: \"Rewards.VoteReward\", account: { id_eq: $subspaceAccount } }\n      limit: 1\n    ) {\n      account {\n        id\n      }\n    }\n    operator: operatorsConnection(\n      first: 1\n      where: { operatorOwner_eq: $subspaceAccount }\n      orderBy: id_ASC\n    ) {\n      totalCount\n    }\n    nominator: nominatorsConnection(\n      first: 1\n      where: { account: { id_eq: $subspaceAccount } }\n      orderBy: id_ASC\n    ) {\n      totalCount\n    }\n  }\n": types.CheckRoleDocument,
```
Comment (EmilFattakhov) [inline explorer/gql/gql.ts:51]: Nice and clean!

File: explorer/src/components/WalletSideKick/GetDiscordRoles.tsx [hunk 3]
```diff
@@ -74,67 +59,95 @@ export const GetDiscordRoles: FC<StakingSummaryProps> = ({ subspaceAccount }) =>
...
+          <Accordion title='Your roles on Discord'>
+            <List>
+              {session?.user?.discord?.vcs.roles.farmer && (
+                <StyledListItem title='You are a Farmer on Discord'>🌾</StyledListItem>
+              )}
+              {session?.user?.discord?.vcs.roles.operator && (
+                <StyledListItem title='You are a Nominator on Discord'>🌐</StyledListItem>
+              )}
+              {session?.user?.discord?.vcs.roles.nominator && (
+                <StyledListItem title='You are a Operator on Discord'>🤝</StyledListItem>
```
Comment (EmilFattakhov) [inline explorer/src/components/WalletSideKick/GetDiscordRoles.tsx:127]:
```suggestion
                <StyledListItem title='You are an Operator on Discord'>🤝</StyledListItem>
```

**Review by EmilFattakhov — DISMISSED** [review]
Looks good, just a few small questions / comments left

**Comment by marc-aurele-besner** [comment]
@EmilFattakhov for your comment in .env.sample

These values should not be on this file, since this PR does not have any GitHub OpenID logic, the GitHub implementation is https://github.com/subspace/astral/pull/578.

To answer your question, atm I only implemented the OpenID connect flow in the subsequent PR, but I did not implement any specific logic, waiting to know what will make sense.

File: explorer/gql/gql.ts (hunk 2 above)
Comment (marc-aurele-besner) [reply explorer/gql/gql.ts:51]: This file is automatically generated when running `yarn codegen` 👍

**Comment by marc-aurele-besner** [comment]
> We just need to get you the role IDs as everything else should already be in place here @marc-aurele-besner?

I already set up the roles in the dev discord server, so as soon as this PR is merged on staging, it should be testable with the dev server.

For production, indeed, we will need to create the "Verified X" role and communicate with me the role ID 👍

**Pull request closed without merging**